In [1]:
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
import numpy 
from sklearn.model_selection import GridSearchCV
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap

Verinin train ve test datası olarak ayrılması

In [2]:
X, y = fetch_openml('mnist_784', version=1, return_X_y=True)
X = X / 255
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)

### Yapay sinir ağı  algoritmasının manuel olarak en iyi parametrelerinin bulunması

MLPClassifier function in Scikit-learn Library is used. I tried different parameters of activation functions, regularization, numbers of hidden units and numbers of hidden layers to find the optimal model for classifying the dataset correctly.

A neural network needs an activation function, hidden layers and regularization parameters. I gave different parameters of each one. The parameters that I gave are shown below:

• Activation Functions : Relu, Tanh, Logistic

• Solver Functions (Optional): Lbfgs, Adam

• Alpha values: (0.01, 0.1, 1.0, 5.0)

• Hidden layer sizes: (1,2)

• Hidden layer units: (10, 50, 100)

#### hidden layer and units [10,10]

In [4]:
#hidden layer and units [10,10]
for this_alpha in [0.01, 0.1, 1.0, 5.0]:
    for this_activation in ['relu', 'tanh', 'logistic']:
        for this_solver in ['lbfgs', 'adam']:
            clf = MLPClassifier(solver=this_solver, 
                        activation = this_activation,
                        alpha = this_alpha,
                        hidden_layer_sizes = [10, 10], 
                        random_state = 0)
            clf.fit(X_train, y_train)
            title = 'Dataset 2: NN classifier, alpha = {:.3f} '.format(this_alpha)
            print("with   |  Solver: ", this_solver , " | Activation: " , this_activation," | Alpha: ", this_alpha)
            print(title)
            print("\nTrain_score: %.3f" % clf.score(X_train, y_train))
            print("Test_score: %.3f\n" % clf.score(X_test, y_test))
            print("------------------------")

with   |  Solver:  lbfgs  | Activation:  relu  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.922
Test_score: 0.911

------------------------


C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  relu  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.967
Test_score: 0.935

------------------------
with   |  Solver:  lbfgs  | Activation:  tanh  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.963
Test_score: 0.929

------------------------


C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  tanh  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.969
Test_score: 0.931

------------------------
with   |  Solver:  lbfgs  | Activation:  logistic  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.955
Test_score: 0.928

------------------------


C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  logistic  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.964
Test_score: 0.936

------------------------
with   |  Solver:  lbfgs  | Activation:  relu  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.923
Test_score: 0.911

------------------------


C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  relu  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.956
Test_score: 0.936

------------------------
with   |  Solver:  lbfgs  | Activation:  tanh  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.963
Test_score: 0.929

------------------------


C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  tanh  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.962
Test_score: 0.939

------------------------
with   |  Solver:  lbfgs  | Activation:  logistic  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.956
Test_score: 0.930

------------------------


C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  logistic  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.946
Test_score: 0.933

------------------------
with   |  Solver:  lbfgs  | Activation:  relu  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.922
Test_score: 0.910

------------------------
with   |  Solver:  adam  | Activation:  relu  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.936
Test_score: 0.929

------------------------
with   |  Solver:  lbfgs  | Activation:  tanh  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.963
Test_score: 0.931

------------------------
with   |  Solver:  adam  | Activation:  tanh  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.940
Test_score: 0.932

------------------------
with   |  Solver:  lbfgs  | Activation:  logistic  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.963
Test_score: 0.937

------------------------

####  hidden layer and units 10

In [5]:
#hidden layer and units 10
for this_alpha in [0.01, 0.1, 1.0, 5.0]:
    for this_activation in ['relu', 'tanh', 'logistic']:
        for this_solver in ['lbfgs', 'adam']:
            clf = MLPClassifier(solver=this_solver, 
                        activation = this_activation,
                        alpha = this_alpha,
                        hidden_layer_sizes = 10, 
                        random_state = 0)
            clf.fit(X_train, y_train)
            title = 'Dataset 2: NN classifier, alpha = {:.3f} '.format(this_alpha)
            print("with   |  Solver: ", this_solver , " | Activation: " , this_activation," | Alpha: ", this_alpha)
            print(title)
            print("\nTrain_score: %.3f" % clf.score(X_train, y_train))
            print("Test_score: %.3f\n" % clf.score(X_test, y_test))

with   |  Solver:  lbfgs  | Activation:  relu  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.918
Test_score: 0.903



C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  relu  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.961
Test_score: 0.935

with   |  Solver:  lbfgs  | Activation:  tanh  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.960
Test_score: 0.927



C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  tanh  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.962
Test_score: 0.927

with   |  Solver:  lbfgs  | Activation:  logistic  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.957
Test_score: 0.927



C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  logistic  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.957
Test_score: 0.934

with   |  Solver:  lbfgs  | Activation:  relu  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.918
Test_score: 0.902

with   |  Solver:  adam  | Activation:  relu  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.954
Test_score: 0.938

with   |  Solver:  lbfgs  | Activation:  tanh  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.959
Test_score: 0.928

with   |  Solver:  adam  | Activation:  tanh  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.956
Test_score: 0.937

with   |  Solver:  lbfgs  | Activation:  logistic  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.957
Test_score: 0.928



C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  logistic  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.936
Test_score: 0.922

with   |  Solver:  lbfgs  | Activation:  relu  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.918
Test_score: 0.902

with   |  Solver:  adam  | Activation:  relu  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.934
Test_score: 0.925

with   |  Solver:  lbfgs  | Activation:  tanh  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.960
Test_score: 0.929

with   |  Solver:  adam  | Activation:  tanh  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.926
Test_score: 0.917

with   |  Solver:  lbfgs  | Activation:  logistic  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.958
Test_score: 0.931

with   |  Solver:  adam  | Activation:  logistic  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.899
Test_score: 0.891

with 

#### hidden layer and units [10,10,10]

In [7]:
#hidden layer and units [10,10,10]
for this_alpha in [0.01, 0.1, 1.0, 5.0]:
    for this_activation in ['relu', 'tanh', 'logistic']:
        for this_solver in ['lbfgs', 'adam']:
            clf = MLPClassifier(solver=this_solver, 
                        activation = this_activation,
                        alpha = this_alpha,
                        hidden_layer_sizes = [10,10,10], 
                        random_state = 0)
            clf.fit(X_train, y_train)
            title = 'Dataset 2: NN classifier, alpha = {:.3f} '.format(this_alpha)
            print("with   |  Solver: ", this_solver , " | Activation: " , this_activation," | Alpha: ", this_alpha)
            print(title)
            print("\nTrain_score: %.3f" % clf.score(X_train, y_train))
            print("Test_score: %.3f\n" % clf.score(X_test, y_test))
            print("------------------------")

with   |  Solver:  lbfgs  | Activation:  relu  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.920
Test_score: 0.907

------------------------
with   |  Solver:  adam  | Activation:  relu  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.969
Test_score: 0.936

------------------------
with   |  Solver:  lbfgs  | Activation:  tanh  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.963
Test_score: 0.932

------------------------
with   |  Solver:  adam  | Activation:  tanh  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.974
Test_score: 0.933

------------------------
with   |  Solver:  lbfgs  | Activation:  logistic  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.917
Test_score: 0.897

------------------------


C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  logistic  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.959
Test_score: 0.928

------------------------
with   |  Solver:  lbfgs  | Activation:  relu  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.923
Test_score: 0.912

------------------------
with   |  Solver:  adam  | Activation:  relu  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.961
Test_score: 0.939

------------------------
with   |  Solver:  lbfgs  | Activation:  tanh  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.963
Test_score: 0.933

------------------------


C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  tanh  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.970
Test_score: 0.942

------------------------
with   |  Solver:  lbfgs  | Activation:  logistic  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.927
Test_score: 0.909

------------------------


C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  logistic  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.948
Test_score: 0.936

------------------------
with   |  Solver:  lbfgs  | Activation:  relu  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.923
Test_score: 0.913

------------------------
with   |  Solver:  adam  | Activation:  relu  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.948
Test_score: 0.936

------------------------
with   |  Solver:  lbfgs  | Activation:  tanh  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.964
Test_score: 0.933

------------------------
with   |  Solver:  adam  | Activation:  tanh  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.950
Test_score: 0.937

------------------------
with   |  Solver:  lbfgs  | Activation:  logistic  | Alpha:  1.0
Dataset 2: NN classifier, alpha = 1.000 

Train_score: 0.934
Test_score: 0.912

------------------------

#### hidden layer and units 50

In [9]:
#hidden layer and units 50
for this_alpha in [0.01, 0.1, 1.0, 5.0]:
    for this_activation in ['relu', 'tanh', 'logistic']:
        for this_solver in ['lbfgs', 'adam']:
            clf = MLPClassifier(solver=this_solver, 
                        activation = this_activation,
                        alpha = this_alpha,
                        hidden_layer_sizes = 50, 
                        random_state = 0)
            clf.fit(X_train, y_train)
            title = 'Dataset 2: NN classifier, alpha = {:.3f} '.format(this_alpha)
            print("with   |  Solver: ", this_solver , " | Activation: " , this_activation," | Alpha: ", this_alpha)
            print(title)
            print("\nTrain_score: %.3f" % clf.score(X_train, y_train))
            print("Test_score: %.3f\n" % clf.score(X_test, y_test))
            print("------------------------")

with   |  Solver:  lbfgs  | Activation:  relu  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 1.000
Test_score: 0.967

------------------------
with   |  Solver:  adam  | Activation:  relu  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 1.000
Test_score: 0.971

------------------------
with   |  Solver:  lbfgs  | Activation:  tanh  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 1.000
Test_score: 0.960

------------------------
with   |  Solver:  adam  | Activation:  tanh  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 1.000
Test_score: 0.970

------------------------
with   |  Solver:  lbfgs  | Activation:  logistic  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 1.000
Test_score: 0.960

------------------------


C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  logistic  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.998
Test_score: 0.973

------------------------
with   |  Solver:  lbfgs  | Activation:  relu  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 1.000
Test_score: 0.968

------------------------
with   |  Solver:  adam  | Activation:  relu  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.992
Test_score: 0.974

------------------------
with   |  Solver:  lbfgs  | Activation:  tanh  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 1.000
Test_score: 0.962

------------------------
with   |  Solver:  adam  | Activation:  tanh  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.991
Test_score: 0.973

------------------------
with   |  Solver:  lbfgs  | Activation:  logistic  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 1.000
Test_score: 0.961

-----------------------

#### hidden layer and units 100

In [14]:
#hidden layer and units 100
for this_alpha in [0.01, 0.1, 1.0, 5.0]:
    for this_activation in ['relu', 'tanh', 'logistic']:
        for this_solver in ['lbfgs', 'adam']:
            clf = MLPClassifier(solver=this_solver, 
                        activation = this_activation,
                        alpha = this_alpha,
                        hidden_layer_sizes = 100, 
                        random_state = 0)
            clf.fit(X_train, y_train)
            title = 'Dataset 2: NN classifier, alpha = {:.3f} '.format(this_alpha)
            print("with   |  Solver: ", this_solver , " | Activation: " , this_activation," | Alpha: ", this_alpha)
            print(title)
            print("\nTrain_score: %.3f" % clf.score(X_train, y_train))
            print("Test_score: %.3f\n" % clf.score(X_test, y_test))
            print("------------------------")

with   |  Solver:  lbfgs  | Activation:  relu  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 1.000
Test_score: 0.974

------------------------
with   |  Solver:  adam  | Activation:  relu  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.998
Test_score: 0.976

------------------------
with   |  Solver:  lbfgs  | Activation:  tanh  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 1.000
Test_score: 0.966

------------------------
with   |  Solver:  adam  | Activation:  tanh  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 1.000
Test_score: 0.976

------------------------
with   |  Solver:  lbfgs  | Activation:  logistic  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 1.000
Test_score: 0.969

------------------------
with   |  Solver:  adam  | Activation:  logistic  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.999
Test_score: 0.978

------------------

#### hidden layer and units [10,100]

In [4]:
#hidden layer and units [10,100]
for this_alpha in [0.01, 0.1, 1.0, 5.0]:
    for this_activation in ['relu', 'tanh', 'logistic']:
        for this_solver in ['lbfgs', 'adam']:
            clf = MLPClassifier(solver=this_solver, 
                        activation = this_activation,
                        alpha = this_alpha,
                        hidden_layer_sizes = [10, 100], 
                        random_state = 0)
            clf.fit(X_train, y_train)
            title = 'Dataset 2: NN classifier, alpha = {:.3f} '.format(this_alpha)
            print("with   |  Solver: ", this_solver , " | Activation: " , this_activation," | Alpha: ", this_alpha)
            print(title)
            print("\nTrain_score: %.3f" % clf.score(X_train, y_train))
            print("Test_score: %.3f\n" % clf.score(X_test, y_test))
            print("------------------------")

with   |  Solver:  lbfgs  | Activation:  relu  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.948
Test_score: 0.938

------------------------


C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  relu  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.986
Test_score: 0.946

------------------------
with   |  Solver:  lbfgs  | Activation:  tanh  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.968
Test_score: 0.936

------------------------


C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  tanh  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.984
Test_score: 0.942

------------------------
with   |  Solver:  lbfgs  | Activation:  logistic  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.950
Test_score: 0.930

------------------------


C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


with   |  Solver:  adam  | Activation:  logistic  | Alpha:  0.01
Dataset 2: NN classifier, alpha = 0.010 

Train_score: 0.967
Test_score: 0.941

------------------------
with   |  Solver:  lbfgs  | Activation:  relu  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.947
Test_score: 0.935

------------------------
with   |  Solver:  adam  | Activation:  relu  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.976
Test_score: 0.957

------------------------
with   |  Solver:  lbfgs  | Activation:  tanh  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.968
Test_score: 0.936

------------------------
with   |  Solver:  adam  | Activation:  tanh  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.965
Test_score: 0.943

------------------------
with   |  Solver:  lbfgs  | Activation:  logistic  | Alpha:  0.1
Dataset 2: NN classifier, alpha = 0.100 

Train_score: 0.951
Test_score: 0.930

-----------------------

Considering these results, looking at the overfit and underfit states between train and test scores, I decided to take it for the best score. We can list the parameters that have the best test score as follows:

**Solver: adam, Activation: relu, Alpha: 0.1, Hidden_layer_sizes: 100**

Train Score: 0.994

Test Score: 0.978

# using GridSerchCV

In [3]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
parameters = {'activation': ['tanh', 'relu', 'logistic'],
              'alpha': [1.0, 0.5, 0.1, 0.01], 
              'hidden_layer_sizes': (10,[10,10],[10,10,10],50,100,[10,100])} 
clf_grid = GridSearchCV(MLPClassifier(), parameters, n_jobs=-1)
clf_grid.fit(X_train, y_train)

C:\Users\Merve\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)


GridSearchCV(cv='warn', error_score='raise-deprecating',
       estimator=MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(100,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=None, shuffle=True, solver='adam', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False),
       fit_params=None, iid='warn', n_jobs=-1,
       param_grid={'activation': ['tanh', 'relu', 'logistic'], 'alpha': [1.0, 0.5, 0.1, 0.01], 'hidden_layer_sizes': (10, [10, 10], [10, 10, 10], 50, 100, [10, 100])},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [11]:
best_score = clf_grid.best_score_

In [8]:
best_parameters = clf_grid.best_params_

In [12]:
best_score

0.9759047619047619

In [13]:
best_parameters

{'activation': 'relu', 'alpha': 0.1, 'hidden_layer_sizes': 100}

## classification with the best parameters 

In [7]:
clf = MLPClassifier(activation= 'relu', alpha= 0.1, hidden_layer_sizes=100).fit(X_train, y_train)
clf.score(X_train, y_train)

0.9950285714285714

In [8]:
clf.score(X_test, y_test)

0.9767428571428571